In [1]:
import numpy as np
from math import factorial
def arrangements(n, k):
    return int(factorial(n) / factorial(n - k))
def permutations(n):
    return int(factorial(n))
def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

### Задача 1
Контрольная работа состоит из пяти вопросов. На каждый вопрос приведено четыре варианта ответа, один из которых правильный. Составьте закон распределения числа правильных ответов при простом угадывании. Найдите M(X), D(X).

#### Решение:
Поскольку на каждый вопрос есть 4 варианта ответов, из которых правилен лишь один, то вероятность правильного ответа на каждый вопрос будет для всех вопросов одинакова и равна 1/4. Вероятность неправильного ответа для всех вопросов также будет одинакова и равна 3/4.
Возможны следующие варианты угадывания правильных ответов: 0 правильных ответов, 1 из 5 правильных ответов, 2 из 5, 3, 4 и 5 правильных ответов. Вероятность каждого варианта рассчитаем по формуле Бернулли, где n = 5, k меняется от 0 до 5, p = 0.25, q = 0.75.
Посчитаем вероятность каждого значения:

$$P(X = 0) = C_n^k p^k (1 - p)^{n - k} = C_5^0 \cdot 0.25^0 \cdot 0.75^5 = \frac{5!}{0! \: (5-0)!} \cdot 0.25^0 \cdot 0.75^5 = 1 \cdot 1 \cdot 0.2373046875 = 0.2373046875$$

$$P(X = 1) = C_5^1 \cdot 0.25^1 \cdot 0.75^4 = \frac{5!}{1! \: 4!} \cdot 0.25^1 \cdot 0.75^4 = 0.3955078125$$

$$P(X = 2) = C_5^2 \cdot 0.25^2 \cdot 0.75^3 = \frac{5!}{2! \: 3!} \cdot 0.25^2 \cdot 0.75^3 = 0.52734375$$

$$P(X = 3) = C_5^3 \cdot 0.25^3 \cdot 0.75^2 = \frac{5!}{3! \: 2!} \cdot 0.25^3 \cdot 0.75^2 = 0.17578125$$

$$P(X = 4) = C_5^4 \cdot 0.25^4 \cdot 0.75^1 = \frac{5!}{4! \: 1!} \cdot 0.25^4 \cdot 0.75^1 = 0.0146484375$$

$$P(X = 5) = C_5^5 \cdot 0.25^5 \cdot 0.75^0 = \frac{5!}{5! \: 0!} \cdot 0.25^5 \cdot 0.75^0 = 0.0009765625$$

Итак, закон распределения такой случайной величины:
<table border="3">
    <tr>
        <th>$x$</th>
        <td>0</td>
        <td>1</td>
        <td>2</td>
        <td>3</td>
        <td>4</td>
        <td>5</td>       
</tr>
    <tr>
        <th>$P(X=x)$</th>
        <td>0.237</td>
        <td>0.396</td>
        <td>0.527</td>
        <td>0.176</td>
        <td>0.015</td>
        <td>0.001</td>        
    </tr>
</table>

Посчитаем математическое ожидание:

In [2]:
x_values = np.array([0, 1, 2, 3, 4, 5])
x_probabilities = np.array([0.2373046875, 0.3955078125, 0.52734375, 
                            0.17578125, 0.0146484375, 0.0009765625])
m = x_values.dot(x_probabilities)
m

2.041015625

In [3]:
y_values = x_values - m
z_values = y_values ** 2
d = z_values.dot(x_probabilities)
d

1.6444779932498932

#### Ответ:
<table border="3">
    <tr>
        <th>$x$</th>
        <td>0</td>
        <td>1</td>
        <td>2</td>
        <td>3</td>
        <td>4</td>
        <td>5</td>       
</tr>
    <tr>
        <th>$P(X=x)$</th>
        <td>0.237</td>
        <td>0.396</td>
        <td>0.527</td>
        <td>0.176</td>
        <td>0.015</td>
        <td>0.001</td>        
    </tr>
</table>

M(X) = 2.041015625

D(X) = 1.6444779932498932

### Задача 2
Вероятность попадания в цель при одном выстреле равна 0.01. Какова вероятность того, что число попаданий при 200 выстрелах будет не менее 5 и не более 10?

#### Решение:
p = 0.01, q = 0.99, n = 200, k = [5 , 10]

$$P(X = 5) = C_{200}^5 \cdot 0.01^5 \cdot 0.99^{195} = \frac{200!}{5! \: 195!} \cdot 0.01^5 \cdot 0.99^{195} = 0.3955078125$$

$$P(X = 6) = C_{200}^6 \cdot 0.01^6 \cdot 0.99^{194} = \frac{200!}{6! \: 194!} \cdot 0.01^6 \cdot 0.99^{194} = 0.3955078125$$

$$P(X = 7) = C_{200}^7 \cdot 0.01^7 \cdot 0.99^{193} = \frac{200!}{7! \: 193!} \cdot 0.01^7 \cdot 0.99^{193} = 0.3955078125$$

$$P(X = 8) = C_{200}^8 \cdot 0.01^8 \cdot 0.99^{192} = \frac{200!}{8! \: 192!} \cdot 0.01^8 \cdot 0.99^{192} = 0.3955078125$$

$$P(X = 9) = C_{200}^9 \cdot 0.01^9 \cdot 0.99^{191} = \frac{200!}{9! \: 191!} \cdot 0.01^9 \cdot 0.99^{191} = 0.3955078125$$

$$P(X = 10) = C_{200}^{10} \cdot 0.01^{10} \cdot 0.99^{190} = \frac{200!}{10! \: 190!} \cdot 0.01^{10} \cdot 0.99^{190} = 0.3955078125$$


In [4]:
n=200
sum(combinations(n, k)*0.01**k*0.99**(n-k) for k in range(5, 11))

0.051739381860374606

#### Ответ:
0.0517

### Задача 3
В условиях предыдущей задачи сколько выстрелов нужно сделать чтобы быть уверенным с вероятностью 0.9, что хотя бы 10 раз будет совершено попадание?

_Подсказка_. Как посчитать P(X >= k) для величины X, имеющей биномиальное распределение? Здесь предстоит немножко покодить.

#### Решение:


In [10]:
n = 10
while p <= 0.9:
    p = sum(combinations(n, k)*0.01**k*0.99**(n-k) for k in range(10, n+1)) 
    n +=1
print(n)

OverflowError: integer division result too large for a float

#### Ответ:
